src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -29,20 +35,66 @@ public SolutionAssetProvider(HostWorkspaceServices services)
+            }
+            else
+            {
+                assetMap = await assetStorage.GetAssetsAsync(scopeId, checksums, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC pipe.
+            // Run two tasks - the first synchronously writes to a local pipe and the second asynchronosly transfers the data to the RPC pipe.
+            //
+            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but might result in more
+            // (non-contiguous) memory allocated for the underlying buffers. The amount of memory is bounded by the total size of the serialized assets.
+            var localPipe = new Pipe(RemoteHostAssetSerialization.PipeOptionsWithUnlimitedWriterBuffer);
 
-                var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
-                var serializer = _services.GetRequiredService<ISerializerService>();
-                await RemoteHostAssetSerialization.WriteDataAsync(writer, assetStorage, serializer, scopeId, checksums, cancellationToken).ConfigureAwait(false);
-            }, cancellationToken);
+            Task.Run(() =>
+            {
+                try
+                {
+                    using var stream = localPipe.Writer.AsStream(leaveOpen: false);
+                    using var writer = new ObjectWriter(stream, leaveOpen: false, cancellationToken);
+                    RemoteHostAssetSerialization.WriteData(writer, singleAsset, assetMap, serializer, scopeId, checksums, cancellationToken);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e, cancellationToken))
+                {
+                    // no-op
+                }
+            }, cancellationToken).Forget();
+
+            async Task CopyPipeData()
+            {
+                Exception? exception = null;
+                try
+                {
+                    await localPipe.Reader.CopyToAsync(pipeWriter, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    FatalError.ReportWithoutCrashUnlessCanceled(e, cancellationToken);
+                    exception = e;
+                }
+                finally
+                {
+                    await localPipe.Reader.CompleteAsync(exception).ConfigureAwait(false);
+                    await pipeWriter.CompleteAsync(exception).ConfigureAwait(false);
+                }
+            }
 
-            return default;
+            // Complete RPC one we send the initial piece of data and start waiting for the writer to send more,
+            // so the client can start reading from the stream.
+            CopyPipeData().Forget();
```

Comment:
i don't quite get why this async opearion is one that is forgotten.  i would assume teh caller of GetAssetsAsync would expect the writer to be complete once their async op finishes.   but that doesn't seem to be the case.  instead, it looks like only some amount of the data may have been written into the pipe.

is it because the caller waits on the pipe being 'completed'?  if so... these mtehods def fee complex in that there are two source of async completion callers need to be familiar with.

Reply:
> can track completion by virtue of the pipe itself.

Yes, that's correct.